### ISSUE TITLE
Possible Inaccurate Date Range

### ISSUE THREAD

**Issue by NateSchwarz** [formtools/module-data_visualization#21]
Hi Ben,

I'm having an issue with the date range accurately pulling in data. I'm using the latest version of form tools and data vis with the new date range abilities on field charts we messaged about a month or so ago.

On the field charts, the date range (specifically year-to-date and previous 30 days) seem to be doing something in terms of pulling data, but I'm noticing some discrepancies. The year-to-date seems to be working pulling in an amount of submissions that makes sense with all my filtering. However, the previous 30 day range is pulling in a very similar amount of submissions. To clarify, the two definitely shouldn't be pulling anywhere close to the same number of submissions.

If this makes sense, there are about 4000 submissions over the past two-ish years. (Just as an example) I'm filtering how many people chose the color blue as an option. I'm using a view which parses out all submissions with the option "blue" then I'm then trying to use the YTD and previous 30 days filters to see how many people selected that option over those two time intervals. 

I don't know if I'm missing something, or I'm not setting up the visualization correctly, but it looks like the time filter is doing something because the numbers do change slightly (a few submissions here and there) between the two time filters, just not in anyway that makes sense or is accurate.

I dug around the code and was able to figure out (at least I think) that the filters are checking the submission date to do their filtering. So I'm wondering if there is bug somewhere? 

I could be missing something blatantly obvious and apologize if I am. I'm just wondering what could be causing inaccuracy in the filtering.

**Comment by benkeen** [comment]
Hey @NateSchwarz, it looks like the query isn't factoring in whether the submission is finalized or not, which is clearly incorrect. Is this an API form where you're seeing this problem?

**Comment by NateSchwarz** [comment]
Hi Ben,

No I'm not running any APIs for it. It's just a form builder form run on page using an iframe so nothing special.

**Comment by NateSchwarz** [comment]
Yep, returned the correct current time and date

**Comment by benkeen** [comment]
Hmph! I gotta say, nothing is standing out as incorrect besides the is_finalized thing mentioned above.
- Are you viewing the field chart as an administrator or client? 
- Does the view you're in contain any filters? (standard / client)

**Comment by benkeen** [comment]
Let's do a little brain surgery and get it to output to the logs the exact SQL query it's using.

On line 165 of `modules/data-visualization/code/FieldCharts.class.php` you'll see this chunk:

```php
        if ($ignore_empty_fields == "yes") {
            $db->query("
                SELECT $col_name as field_value, count(*) as count
                FROM   {PREFIX}form_{$form_id}
                WHERE  1=1
                	   AND $col_name IS NOT NULL AND TRIM($col_name) != ''
                       $filter_where_clause
                       $date_range_clause
                GROUP BY field_value
            ");
        } else {
```

Change that to:

```
        if ($ignore_empty_fields == "yes") {
            $query = "
                SELECT $col_name as field_value, count(*) as count
                FROM   {PREFIX}form_{$form_id}
                WHERE  1=1
                	   AND $col_name IS NOT NULL AND TRIM($col_name) != ''
                       $filter_where_clause
                       $date_range_clause
                GROUP BY field_value
            ";
            $db->query($query);
            error_log($query);
        } else {
```

That'll now log the exact query that it's using to your PHP file. If it's not sensitive, mind posting it here? Anything suspicious in the query - wrong dates being applied, or extra fields being checked against?

**Comment by NateSchwarz** [comment]
Hey Ben,

I gave that a try, where exactly am I looking for the logging? I assumed the error log but nothing was updated.

**Comment by benkeen** [comment]
Yeah, the PHP logs should be where you see it. That's strange - that location you edited was for Field Charts with ignore empty fields = "yes" and would execute when it's not cached... 

But actually, I think it might be moot. I realized that other fix I made to this repo may also fix your scenario. Form Builder forms _also_ use the `is_finalized` setting so you may be affected by that bug. Basically that setting is used to mark a submission in completed or not, so form builder forms flag records as not finalized until the user submits the form. As such, the Field Chart may well be showing you invalid data, 

Try this: change that one line in the SQL statement above from `WHERE  1=1` to `WHERE is_finalized = 'yes'`. Also make the same change a few lines lower in the next DB query. 

Let me know how your chart looks after that change.

**Comment by NateSchwarz** [comment]
I tried that and the charts aren't looking any different or outputting any different data unfortunately.

**Comment by NateSchwarz** [comment]
Hey Ben, 

I just wanted to let you know, I found an anomaly with the dates in my data after I did the last import a few nights ago. There are about 5000 records I had to transition from an old database to the new one I'm working with, and a chunk of the dates somehow got corrupted between excel and phpmyadmin. I caught it the other night when I noticed I had entries from 2031 which led me to believe as I kept exporting and importing the problem compounded on itself, and I didn't realise until it was so severe. 

I'm still working with the dates to make sure they are correct, but the problem may have not been in the code after all.
